### PR TITLE
General clean-up of the single stream connections code

### PR DESCRIPTION
**Pull request by tomaka**
This PR is a general clean-up of the single stream connection code. It fixes many small issues, and also handles the shutdown phase more properly.
We now properly handle incoming GoAway frames, and have the possibility to send some.

Can be reviewed commit by commit.
I understand that this code is generally extremely complicated, but I don't really have a good solution for this. Networking code *is* extremely complicated, both in terms of logic and implementation.
Maybe a solution is tests, such as https://github.com/paritytech/smoldot/pull/2323, but at the moment the code isn't even robust enough to have tests.

### CODE REVIEW

File: src/libp2p/connection/established/single_stream.rs [hunk 1]
```diff
@@ -258,12 +266,17 @@ where
                 .map_err(Error::Yamux)?;
             self.inner.yamux = yamux_decode.yamux;
 
-            // TODO: it is possible that the yamux reading is blocked on writing
+            // If bytes_read is 0 and detail is None, then Yamux can't do anything more. On the
```
Comment (melekes) [inline src/libp2p/connection/established/single_stream.rs:292]: I don't see how this comment adds anything. it just duplicates the code basically

File: src/libp2p/connection/yamux.rs [hunk 2]
```diff
@@ -1564,6 +1604,33 @@ impl<'a, T> ExtractOut<'a, T> {
                 }
 
                 Outgoing::Idle => {
+                    // Send a "GoAway" frame if demanded.
+                    if let OutgoingGoAway::Required(code) = self.yamux.outgoing_goaway {
+                        let mut header = arrayvec::ArrayVec::new();
+                        header.push(0);
```
Comment (melekes) [inline src/libp2p/connection/yamux.rs:1739]: what `0` is representing here?

Comment (tomaka) [reply src/libp2p/connection/yamux.rs:1739]: It's a Yamux version number. See https://github.com/hashicorp/yamux/blob/master/spec.md#version-field

There are several snippets of code that generate a header, and it's true that it wouldn't be a bad idea to move them to the `header` module for example, but I also don't think it's a big deal.

File: src/libp2p/connection/established/single_stream.rs (hunk 1 above)
Comment (tomaka) [reply src/libp2p/connection/established/single_stream.rs:292]: It's true that it's repeating a piece of information of the Yamux API.
But IMO, the fact that we have to loop again if Yamux didn't process all the data or returned an event is a bit subtle (especially the "or" here, so for example if Yamux processed all the data and returned an event we have to loop again), and is worth repeating for clarity.

**Review by melekes — APPROVED** [review]
👍

**Pull request merged**